### PR TITLE
Add minimal CUDA/MPI test pipeline 

### PR DESCRIPTION
**Pull request by nefrathenrici**
This PR replaces the old pipeline with a new one which runs a basic CUDA/MPI functionality test on central, clima, and derecho. I originally used a generated pipeline (still included for now) but it isn't really needed if we only have 3 clusters to test.

It currently fails with a generic MPI error on central, I am not sure why yet. I have tried srun and mpiexec, both lead to separate errors.

https://buildkite.com/clima/test-cuda-mpi/builds/17#_

### CODE REVIEW

File: .buildkite/pipeline.sh
Comment (Sbozzolo) [inline .buildkite/pipeline.sh:1]: I personally find this much harder to understand than a non-dynamic pipeline.

I don't think we benefit too much from making this dynamic, so I'd suggest we convert this into a normal yml file

File: .buildkite/test_cuda_mpi.jl [hunk 1]
```diff
@@ -0,0 +1,79 @@
+#= To run:
+nsys profile --trace=nvtx,cuda,mpi --output=report_%q{PMI_RANK} --force-overwrite true julia --project bin/test_cuda_mpi.jl
```
Comment (Sbozzolo) [inline .buildkite/test_cuda_mpi.jl:2]: Don't you need to add an `mpiexec` or something like that?

File: .buildkite/test_cuda_mpi.jl [hunk 2]
```diff
@@ -0,0 +1,79 @@
+#= To run:
+nsys profile --trace=nvtx,cuda,mpi --output=report_%q{PMI_RANK} --force-overwrite true julia --project bin/test_cuda_mpi.jl 
+nsys stats --report cuda_gpu_trace report_0.nsys-rep --force-export true
+=#
+
```
Comment (Sbozzolo) [inline .buildkite/test_cuda_mpi.jl:5]: Can you add a description of the tests performed?

File: .buildkite/test_cuda_mpi.jl [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+- `test_gpu_visbility`: Test access to at least one GPU and perform a minimal computation
+- `test_mpi`: Test a send/receive between arrays on MPI ranks. This is used to
+test transfers between CPU-allocated Arrays and GPU-allocated CuArrays.
+
+For performance, is important to verify that transfers between CuArrays on different GPUs 
+are peer-to-peer transfers, which is done by profiling the test with Nsight Systems.
+
+To run this test, you need 2 CUDA devices, CUDA-aware MPI, and Nsight Systems.
+1. mpirun -n 2 nsys profile --trace=nvtx,cuda,mpi julia --project bin/test_cuda_mpi.jl 
+2. nsys stats --report cuda_gpu_trace report1.nsys-rep
```
Comment (Sbozzolo) [inline .buildkite/test_cuda_mpi.jl:13]: Could you describe the expected output for a successful test?

Comment (nefrathenrici) [reply .buildkite/test_cuda_mpi.jl]: Here is the new comment in the test file:

#= Test basic CUDA and CUDA-aware MPI functionality

Main testing functions:

- `test_gpu_visbility`: Test access to at least one GPU and perform a minimal computation
- `test_mpi`: Test a send/receive between arrays on MPI ranks. This is used to
test transfers between CPU-allocated Arrays and GPU-allocated CuArrays.

For performance, is important to verify that transfers between CuArrays on different GPUs 
are peer-to-peer transfers, which is done by profiling the test with Nsight Systems.

To run this test, you need 2 CUDA devices, CUDA-aware MPI, and Nsight Systems.

First, run the test with two MPI ranks, profiling with Nsight:

    `mpirun -n 2 nsys profile --trace=nvtx,cuda,mpi julia --project bin/test_cuda_mpi.jl`

If this runs successfully, the test prints MPI and CUDA version info and the following output:

```
Testing GPU Visibility
Basic GPU computation successful
Testing MPI with Array
Transfer successful
Testing MPI with CuArray
Transfer successful
All tests completed!
```

Once the profile has been generated, analyze the data to determine if a 
peer-to-peer transfer occured between the GPUs:

`nsys stats --report cuda_gpu_trace report1.nsys-rep | grep grep -E "memcpy (Peer-to-Peer|PtoP)"`

You should see `[CUDA memcpy Peer-to-Peer]` if the peer-to-peer transfer was successful.
If you don't see this, you can view the full profile analysis using

`nsys stats --report cuda_gpu_trace report1.nsys-rep`
=#

**Review by Sbozzolo — APPROVED** [review]
Thank you!

**Pull request merged**